### PR TITLE
[charts/redis-ha] Add network policy

### LINKED ISSUE #83: [chart/redis-ha] Implement network policies in the chart
**Is your feature request related to a problem? Please describe.**
We use network policy in our clusters and the Redis don't work if Network Policy are defined inside the namespace where Redis is running.

**Describe the solution you'd like**
I would like the to have proper Network Policies to make Redis work in case of a policy like this:

```
---
kind: NetworkPolicy
apiVersion: networking.k8s.io/v1
metadata:
  name: default.deny-all-ingress-traffic
  namespace: default
spec:
  podSelector: {}
  policyTypes:
    - Ingress
```

**Describe alternatives you've considered**
There is no alternative

**Additional context**
Ref: https://kubernetes.io/docs/concepts/services-networking/network-policies/

### LINKED ISSUE #106: [chart/redis_ha][REQUEST] Small improvements
I have a couple of small requests, most of which I could make a PR myself on, but I'd like to hear from maintainers first. :)

1. Our K8s cluster requires containers to define a `readinessProbe`. None of the containers in this chart do that right now. The readiness probe could just be the same `ping` that the `livenessProbe` does. I had to fix that on my fork and I'd love to make a PR on that.

2. Network policy - While those are not explicitly enforced they are mandated for our production environment. I can of course create one outside of this chart, but it'd be nice to have it as part of it. The chart could define a label that pods allowed to connect Redis/HAProxy must have in order to be allowed. I am however unsure exactly how the HAProxy <-> Redis connectivity works so not 100% on how to define the policy correctly (In our deployment we skipped HAProxy).

3. I have noticed that the cluster is in an incorrect state if you configure it with:
```yaml
redis:
    tlsReplication: true
    tlsPort: 6380
    port: 6379
  sentinel:
    tlsReplication: true
    tlsPort: 26380
    port: 26379
    auth: true
    existingSecret: <redacted>
    authKey: redis-password
    config:
      down-after-milliseconds: 5000
      failover-timeout: 10000
      parallel-syncs: 5
      maxclients: 10000
  auth: true
  existingSecret: <redacted>
  authKey: redis-password
  tls:
    secretName: <redacted>
    certFile: redis-crt
    keyFile: redis-key
    caCertFile: root-ca
```

The problem occurs when both `port` and `tlsPort` are defined. The Sentinels attempt to communicate with each other on the non-TLS port and that does not succeed. After `down-after-milliseconds` time all Sentinels assume the other Sentinels are gone. Thus they can never reach a quorum and perform a successful failover. After I reconfigured with:

```yaml
  redis:
    tlsReplication: true
    tlsPort: 6380
    port: 0
  sentinel:
    tlsReplication: true
    tlsPort: 26380
    port: 0
```

Things started working properly. I'm not sure what is the intention here, but at least some documentation warning that such misconfiguration is possible would be nice.

### PR DESCRIPTION
**Pull request by jdheyburn**
#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #83 
  - fixes #106 

#### Special notes for your reviewer:
- Included a minimal set of rules by default to avoid initial complexity.
- Does not include ingress rule for exporter, since this can vary quite a lot in environments. The same can be achieved with an `ingressRule` that the user can define.
- I don't have much context on haproxy (I do not use it in this chart) so I can't contribute to it there. However I imagine would be handled in another template file similar to this one. 
- Did not include a DNS rule either - again this can be defined above or perhaps consumers will have a global allow DNS rule.
- `ports` for an ingress rule is optional (defaults to Redis and Sentinel ports)
- `ports` for an egress rule is required (no sensible default assumed)

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://github.com/helm/charts/blob/master/CONTRIBUTING.md#sign-your-work) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md
- [x] Title of the PR starts with chart name (e.g. `[stable/mychartname]`)

### CODE REVIEW

**Comment by sean-mcgimpsey** [comment]
Looks good - though just wondering if we are able to make the ingress/egress selectors templatable?

**Comment by jdheyburn** [comment]
@sean-mcgimpsey It's a fair suggestion, I just worry about making something _too opinionated_ in place of flexibility.

Could you suggest an example, and then other readers can comment from there? Thanks! 👍🏻

**Comment by sean-mcgimpsey** [comment]
Hey - sorry for the delay. here would be my suggested changes... 

Within `redis-ha-network-policy.yaml` 

```
---# line 37
{{-  range $rule := .Values.networkPolicy.egressRules }}
    - to:
{{ (tpl (toYaml $rule.selectors) $)| indent 9 }} # note the use of the tpl function
---
```

```
--- # line 54
{{-  range $rule := .Values.networkPolicy.ingressRules }}
    - from:
{{ (tpl (toYaml $rule.selectors) $) | indent 7 }} # note the use of the tpl function
---
```



This would give us the ability to do something like this within the `values.yaml`

```
  networkPolicy:
    ## whether NetworkPolicy for Redis StatefulSets should be created
    ## when enabled, inter-Redis connectivity is created
    enabled: true
    ingressRules:
      - selectors:
        - podSelector:
            matchLabels:
                release: |-
                  {{- .Release.Name }}
                app: |-
                  {{- template "redis-ha.fullname" . }}
```
Essentially it would allow the use of helpers, which gives more flexibility in my opinion.

And of course the below would also work (standard values)
```
  networkPolicy:
    ## whether NetworkPolicy for Redis StatefulSets should be created
    ## when enabled, inter-Redis connectivity is created
    enabled: true
    ingressRules:
      - selectors:
        - podSelector:
            matchLabels:
                release: myrelease
                app: testapp
```

Hope that makes sense. 
Thanks

**Comment by jdheyburn** [comment]
My apologies @sean-mcgimpsey , I misunderstood the ask. Thanks for providing an example, I didn't know that was possible (TIL!) - I've made that change now and tested on my side but would appreciate if you could pull the branch and give it a test too - cheers!

**Comment by sean-mcgimpsey** [comment]
This changed looks good to me - i've added your fork as a helm repo and used it as the url for my redis dependency - the use of tpl looks to be working as expected.

**Review by DandyDeveloper — COMMENTED** [review]
@jdheyburn Looks great, sorry for the delay. Being pulled all over the place recently.

I just have the above comment.

File: charts/redis-ha/templates/redis-ha-network-policy.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    - from:
+{{ (tpl (toYaml $rule.selectors) $) | indent 7 }}
+      ports:
+{{- if $rule.ports }}
+{{ toYaml $rule.ports | indent 7 }}
+{{- else }}
+        - port: {{ $root.Values.redis.port }}
+          protocol: TCP
+        - port: {{ $root.Values.sentinel.port }}
+          protocol: TCP
```
Comment (DandyDeveloper) [inline charts/redis-ha/templates/redis-ha-network-policy.yaml:64]: Don't these need to be defaults? If someone changes this, they lose these ports.

**Review by DandyDeveloper — COMMENTED** [review]
@jdheyburn Looks great, sorry for the delay. Being pulled all over the place recently.

I just have the above comment.

**Comment by DandyDeveloper** [comment]
@jdheyburn Added a CI fix, running through it again now. Will merge on success.

**Review by DandyDeveloper — DISMISSED** [review]

**Review by DandyDeveloper — APPROVED** [review]

File: charts/redis-ha/templates/redis-ha-network-policy.yaml (hunk 1 above)
Comment (jdheyburn) [reply charts/redis-ha/templates/redis-ha-network-policy.yaml:64]: Hey thanks for the comment and sorry for delay, I'm not sure of what the alternative is (am relatively new to NP), would you mind expanding your point?

**Pull request merged**